pkg/cvo/status.go

```diff
@@ -413,6 +420,41 @@ func updateClusterVersionStatus(cvStatus *configv1.ClusterVersionStatus, status
 	}
 }
 
+// filterErrorForFailingCondition filters out update errors based on the given
+// updateEffect. If the err matches the given update effect, nil is
+// returned. If the err has the reason MultipleErrors, its immediate nested
+// errors are filtered out and the error is recreated. Otherwise, the original
+// err is returned.
+func filterErrorForFailingCondition(err error, updateEffect payload.UpdateEffectType) error {
+	if uErr, ok := err.(*payload.UpdateError); ok {
+		if uErr == nil || uErr.UpdateEffect == payload.UpdateEffectNone {
+			return nil
+		}
+		if uErr.Reason == "MultipleErrors" {
```

Comment:
I don't think we need this `Reason` check; can't we `filterOutUpdateErrors(filtered, updateEffect)` for any reason, to avoid relying on logic like "the input error will use `MultipleErrors` if there's update-effect skew between its children"?

Reply:
Not even necessarily update-effect skew. The test case `condensed UpdateError is not modified` showcases this issue to some extent. Let me elaborate on a different test. Let's see a `condenseClusterOperators` test. https://github.com/openshift/cluster-version-operator/blob/99c3a377cd9c88644e3c71097f3d073dc230e7cd/pkg/cvo/sync_worker_test.go#L336-L355

We can have multiple nested errors of various update effects and have the root error as `ClusterOperatorsDegraded`. We can filter out the nested errors for all `Reasons` but we don't know how to recreate the root error message for all `Reasons` if don't keep this information **somewhere**.

We know how to recreate easily errors that have the reason `MultipleErrors`, and we can easily filter out its immediate nested errors and correctly recreate them. For example, this solves the current bug and any similar issues that arise from the summarized task graph error.

This implementation has limitations. It solves only `MultipleErrors` because we know that the `MultipleErrors` contain multiple errors of various effects, we can easily filter them, and recreate the root error again.

> to avoid relying on logic like "the input error will use MultipleErrors if there's update-effect skew between its children"

The root error may contain any number of nested errors, and the root's message may depend on them and any external factors (same applies to the nested errors). Solving all use-cases would require more changes. In case, this simple fix for a large number of cases is not adequate, and a robust fix for all current and potential future cases would be more appreciated, more work needs to be done.

Reply:
Ah, makes sense as "picking cases where it's easy to produce a fresh root error" for now, thanks for the detailed explanation :bow: